extensions/storages/source/providers/S3.ts

```diff
@@ -0,0 +1,101 @@
+import { type Readable } from 'node:stream'
+import { join } from 'node:path'
+import {
+  S3Client,
+  GetObjectCommand,
+  PutObjectCommand,
+  CopyObjectCommand,
+  ListObjectsV2Command,
+  DeleteObjectsCommand
+} from '@aws-sdk/client-s3'
+import { type Provider } from '../Provider'
+
+const {
+  S3_REGION = '',
+  S3_ENDPOINT = '',
+  S3_ACCESS_KEY = '',
+  S3_SECRET_ACCESS_KEY = '',
+  S3_BUCKET = ''
+} = process.env
+
+export class S3 implements Provider {
+  protected readonly path: string
+  protected readonly bucket = S3_BUCKET
+  protected readonly client: S3Client
+
+  public constructor (url: URL) {
+    if (url.host !== '')
+      throw new Error('URL must not contain host')
+
+    this.path = url.pathname
+    this.client = new S3Client({
+      region: S3_REGION,
+      endpoint: S3_ENDPOINT,
+      credentials: {
+        accessKeyId: S3_ACCESS_KEY,
+        secretAccessKey: S3_SECRET_ACCESS_KEY
+      }
+    })
+  }
+
+  public async get (path: string): Promise<Readable | null> {
+    const key = join(this.path, path)
+
+    try {
+      const fileResponse = await this.client.send(new GetObjectCommand({
+        Bucket: this.bucket,
+        Key: key
+      }))
+
+      /**
+       * Code always runs in NodeJS environment,
+       * so there always will be NodeJS.Readable stream,
+       * there is no need of type guard ??
+       */
+      return fileResponse.Body as Readable
+    } catch (err: any) {
+      if (err?.Code === 'NoSuchKey') return null
+      else throw err
+    }
+  }
+
+  public async put (path: string, filename: string, stream: Readable): Promise<void> {
+    const key = join(this.path, path, filename)
```

Comment:
`join` is using OS-specific separator, this will not work on Windows.

Change to `posix.join`.

Reply:
Fixed